impeller/entity/contents/vertices_contents.cc

```diff
@@ -0,0 +1,99 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "vertices_contents.h"
+
+#include "impeller/entity/contents/content_context.h"
+#include "impeller/entity/entity.h"
+#include "impeller/entity/vertices.frag.h"
+#include "impeller/entity/vertices.vert.h"
+#include "impeller/geometry/color.h"
+#include "impeller/renderer/formats.h"
+#include "impeller/renderer/render_pass.h"
+#include "impeller/renderer/sampler_library.h"
+
+namespace impeller {
+
+VerticesContents::VerticesContents(Vertices vertices)
+    : vertices_(std::move(vertices)){};
+
+VerticesContents::~VerticesContents() = default;
+
+std::optional<Rect> VerticesContents::GetCoverage(const Entity& entity) const {
+  return vertices_.GetTransformedBoundingBox(entity.GetTransformation());
+};
+
+bool VerticesContents::Render(const ContentContext& renderer,
+                              const Entity& entity,
+                              RenderPass& pass) const {
+  using VS = VerticesVertexShader;
+
+  const auto coverage_rect = vertices_.GetBoundingBox();
+
+  if (!coverage_rect.has_value()) {
+    return true;
+  }
+
+  if (coverage_rect->size.IsEmpty()) {
+    return true;
+  }
+
+  VertexBufferBuilder<VS::PerVertexData> vertex_builder;
+  std::vector<Point> points = vertices_.GetPoints();
+  std::vector<uint16_t> indices = vertices_.GetIndices();
+  // TODO: colors are currently unused, must be blended with
+  // paint color based on provided blend mode.
+  std::vector<Color> colors = vertices_.GetColors();
+  VertexMode mode = vertices_.GetMode();
+
+  if (indices.size() == 0) {
+    for (uint i = 0; i < points.size(); i += 1) {
+      VS::PerVertexData data;
+      data.point = points[i];
+      data.vertex_color = color_;
+      vertex_builder.AppendVertex(data);
+    }
+  } else {
+    for (uint i = 0; i < indices.size(); i += 1) {
+      VS::PerVertexData data;
+      data.point = points[indices[i]];
```

Comment:
Rather than using a `VertexBufferBuilder` to unroll the vertices (which also creates/uploads a new incrementing index buffer under the hood), the `VertexBuffer` can be created directly here by:
1. allocating a `DeviceBuffer` big enough to store the indices/vertices
2. copying the vertex and index data directly into the buffer
3. constructing a `VertexBuffer` with index/vertex `BufferView`s into the `DeviceBuffer`.

This is also how the imgui backend works ([allocate transient buffer](https://github.com/flutter/engine/blob/a9918b17ce70d1f9775340f8e87eaa698a1c84eb/impeller/playground/imgui/imgui_impl_impeller.cc#L139-L140), [copy stuff to it](https://github.com/flutter/engine/blob/a9918b17ce70d1f9775340f8e87eaa698a1c84eb/impeller/playground/imgui/imgui_impl_impeller.cc#L173-L186), [create `VertexBuffer`](https://github.com/flutter/engine/blob/a9918b17ce70d1f9775340f8e87eaa698a1c84eb/impeller/playground/imgui/imgui_impl_impeller.cc#L236-L247)).

This optimization is slightly more Rube Goldberg, so totally your call whether you'd like to do it now or maybe later in a followup.

Reply:
I backed this change out for now, because Its actually pretty difficult to debug the errors when targeting android. We also have another potential copy we could elide from the displaylist structure into the impeller vertices structure.